### PR TITLE
log/save_interval based on global step

### LINKED ISSUE #3466: Logging Intervals and Evaluation Intervals
## 🚀 Feature
Logging and validation intervals should be performed using the global step.

### Motivation
Currently, the row_log_interval and val_check_interval do not work as intended. If my batch size is large or my training set size is small, the number of batches per epoch is also small. Row log interval and validation check interval seem to use the steps within the epoch, rather than the global step count. 

This means that logging does not occur if row_log_interval > num batches per epoch. For the validation check interval, the trainer throws an exception.

### Pitch
Both intervals should rely on the global step count  or count steps across epochs to solve the issues mentioned above.

### Alternatives
Before initializing the trainer, I've used the dataloader/dataset to calculate the number of batches per epoch, and then used check_val_every_n_epochs. However this is not ideal if the get_dataloader is baked into the lightning module. It also does not allow for fine grain control over when to run validation.

### Additional context
It might also be nice to have val_check_interval based on max_steps. For example, if we have max_steps=10000 and want to validate ten times through training, we could set val_check_interval=0.1 or equivalently, val_check_interval=1000

### LINKED ISSUE #3506: Tensorboard display nothing
### To Reproduce

I run the code below successfuly, and when i open my tensorboard, scalar display in board, but there's not value on the plot, vert confusing~

#### Code sample

```python
class DNNClassifer(pl.LightningModule):
    def __init__(self):
        super().__init__()

        # network
       
        # others
        self.criterion = MSELoss()
        self.logger_kun = init_logger("dnn", notebook=True)
        

    def forward(self, x):
        x = self.model.forward(x)
        return x
    
    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters())
    
    def training_step(self, batch, batch_idx):
        x, y, data_load_times = batch
        y_pred = self(x)
        
        loss = self.criterion(y_pred.squeeze(), y)
        self.logger.experiment.add_scalar('loss', loss, self.global_step)
        result = pl.TrainResult(minimize=loss)
        result.log("train_loss", loss, on_epoch=True)
        return result
        
    def validation_step(self, batch, batch_idx):
        x, y, data_load_times = batch
        y_pred = self(x)
        loss = self.criterion(y_pred.squeeze(), y)
        
        result = pl.EvalResult(checkpoint_on=loss)
        result.log("valid_loss", loss)
        return result
```
image

### Expected behavior

Display value on tensorboard

### Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/PyTorchLightning/pytorch-lightning/master/tests/collect_env_details.py
# For security purposes, please check the contents of collect_env_details.py before running it.
python collect_env_details.py
```

* CUDA:
	- GPU:
	- available:         False
	- version:           10.2
* Packages:
	- numpy:             1.18.1
	- pyTorch_debug:     False
	- pyTorch_version:   1.6.0
	- pytorch-lightning: 0.9.0
	- tensorboard:       2.2.0
	- tqdm:              4.42.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.6
	- version:           #47~18.04.1-Ubuntu SMP Thu May 7 13:10:50 UTC 2020

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

This issue came up frequently: nothing is getting logged, what's the problem?
**If number of batches in training is < row_log_interval, nothing gets logged.**
This PR will fix this by making row_log_interval based on global_step. 

The current logging behaviour does not change, except in the edge case where epochs are shorter than logging interval. 

Fixes #3466 
Fixes #3506 

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by rohitgr7** [comment]
Global step updates only on the training step right? I checked logging it after every train_epoch_end with batches=50 and getting results:
```
Train epoch end: Epoch: 0 Global Step: 49
Train epoch end: Epoch: 1 Global Step: 100
Train epoch end: Epoch: 2 Global Step: 151
Train epoch end: Epoch: 3 Global Step: 202
```
It is incrementing by an extra +1 after every epoch. Is this expected behavior?

**Comment by awaelchli** [comment]
@rohitgr7 Yes I remember discussing this before with someone, I can't find the issue or PR unfortunately. 
I don't remember what the conclusion was, whether it is a bug or not. The +1 comes from the training_epoch_end, if we log there it creates an extra step. As far as I know, it has always been like this. Ideally, the logs in epoch end should be combined with the last log call of the training loop.

**Comment by rohitgr7** [comment]
> Ideally, the logs in epoch end should be combined with the last log call of the training loop.

Agreed.

@awaelchli Ideally IMO `training_epoch_end` is not doing any leaning step and since we are now logging w.r.t global_step, it should not increment at training_epoch_end.

**Comment by awaelchli** [comment]
so you say we should first fix this global step problem before we can switch to logging on global step? ok, that makes sense.

**Comment by rohitgr7** [comment]
maybe just removing this line will do the job.
https://github.com/PyTorchLightning/pytorch-lightning/blob/a94728c99b68085b861f7384c8ecc163b2d401b0/pytorch_lightning/trainer/connectors/logger_connector.py#L237

**Comment by awaelchli** [comment]
@rohitgr7 I'm trying your suggestion and see if any tests fail #3673

**Comment by Borda** [comment]
> @rohitgr7 I'm trying your suggestion and see if any tests fail #3673

what do you mean by all, we may have the bad sound in some other places already in tests...

**Comment by awaelchli** [comment]
> It is incrementing by an extra +1 after every epoch. Is this expected behavior or a bug?

@rohitgr7 william fixed it, it should be good now and we can trust global_step to be aligned in all epochs.

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/training_loop.py [hunk 1]
```diff
@@ -451,7 +451,7 @@ def on_before_backward(self, batch_idx, optimizer):
 
     def _track_gradient_norm(self, batch_idx):
```
Comment (SkafteNicki) [inline pytorch_lightning/trainer/training_loop.py:452]: Remove `batch_idx` as it is not used anymore?

Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py:452]: thanks, I missed that one

Comment (awaelchli) [reply pytorch_lightning/trainer/training_loop.py]: done!

**Review by rohitgr7 — APPROVED** [review]
LGTM :v:﻿

**Review by Borda — APPROVED** [review]
LGTM

**Comment by williamFalcon** [comment]
awesome fix!

**Pull request merged**